src/server/worker/datum_iterator.go

```diff
@@ -83,6 +89,16 @@ func (d *pfsDatumIterator) Datum() []*Input {
 	return []*Input{d.inputs[d.location]}
 }
 
+func (d *pfsDatumIterator) DatumN(n int) []*Input {
+	if n < d.location {
+		d.Reset()
+	}
+	for d.location != n {
+		d.Next()
```

Comment:
Similar to my other comment repeatedly calling `Next` here seems really inefficient, can't you just set location to the value of `n`?